src/lib.rs

```diff
@@ -81,16 +81,49 @@ fn new_impl(ast: &syn::MacroInput, fields: Option<&[syn::Field]>,
             format!("Constructs a new `{}::{}`.", name, variant),
         ),
     };
+    let attrs = collect_parent_attrs(&ast.attrs);
+    let attrs = quote![#(#attrs),*];
     quote! {
         impl #impl_generics #name #ty_generics #where_clause {
             #[doc = #doc]
+            #attrs
             pub fn #new(#(#args),*) -> Self {
                 #name #qual #inits
             }
         }
     }
 }
 
+fn collect_parent_attrs(attrs: &[syn::Attribute]) -> Vec<syn::Attribute> {
+    fn is_lint(item: &syn::MetaItem) -> bool {
+        if let syn::MetaItem::List(ref ident, _) = *item {
+            match ident.as_ref() {
+                "allow" | "deny" | "forbid" | "warn" => true,
+                _ => false,
+            }
+        } else {
+            false
+        }
+    }
+
+    attrs.iter().filter(|attr| {
+        if is_lint(&attr.value) {
+            true
+        } else if let syn::MetaItem::List(ref ident, ref items) = attr.value {
```

Comment:
if you extract this out into a helper function like `is_lint`, then the closure body could be just `is_lint(...) || is_cfg_attr_lint(...)`